File: plugins/alive_stat/collector.py
"""
alive_stat 数据采集模块。
负责系统资源、进程状态、网络连通性的采集。
"""
import asyncio
import platform
from dataclasses import dataclass
from typing import Optional

import psutil

from .config import config

from ..utils.tools import get_logger

logger = get_logger("alive_stat.collector")


# ================= 数据类 =================

@dataclass
class ResourceUsage:
    cpu_percent: float
    mem_used: int
    mem_total: int
    mem_percent: float
    disk_used: int
    disk_total: int
    disk_percent: float


@dataclass
class ProcessInfo:
    name: str
    running: bool
    mem_bytes: int  # RSS（含子进程），0 表示未运行


@dataclass
class NetworkResult:
    host: str
    reachable: bool
    latency_ms: Optional[float]  # None 表示不可达


@dataclass
class ServerStatus:
    hostname: str
    cpu_model: str
    cpu_cores: int
    uptime: str
    resources: ResourceUsage
    processes: list[ProcessInfo] = None
    network: list[NetworkResult] = None

    def __post_init__(self):
        if self.processes is None:
            self.processes = []
        if self.network is None:
            self.network = []


# ================= 格式化工具 =================

def format_bytes(b: int) -> str:
    """将字节数格式化为人类可读的字符串。"""
    if b < 1024:
        return f"{b}B"
    elif b < 1024 ** 2:
        return f"{b / 1024:.1f}K"
    elif b < 1024 ** 3:
        return f"{b / 1024 ** 2:.1f}M"
    else:
        return f"{b / 1024 ** 3:.1f}G"


# ================= 系统信息 =================

def get_cpu_model() -> str:
    try:
        with open("/proc/cpuinfo", "r") as f:
            for line in f:
                if line.startswith("model name"):
                    return line.split(":", 1)[1].strip()
    except Exception:
        pass
    return platform.processor() or "Unknown"


def get_system_uptime() -> str:
    from .runtime import format_duration
    try:
        boot_time = psutil.boot_time()
        from datetime import datetime
        delta = datetime.now() - datetime.fromtimestamp(boot_time)
        return format_duration(delta)
    except Exception:
        return "N/A"


def get_resources() -> ResourceUsage:
    cpu_percent = psutil.cpu_percent(interval=0.5)
    mem = psutil.virtual_memory()
    disk = psutil.disk_usage("/")
    return ResourceUsage(
        cpu_percent=cpu_percent,
        mem_used=mem.used,
        mem_total=mem.total,
        mem_percent=mem.percent,
        disk_used=disk.used,
        disk_total=disk.total,
        disk_percent=disk.percent,
    )


# ================= 进程采集 =================

def _get_processes_psutil() -> list[ProcessInfo]:
    """
    按配置顺序匹配进程。
    已被前面条目统计过的 PID（含子进程）不会被后面重复计算。
    支持用 '|' 分隔多个关键字（OR 匹配）。
    """
    global_counted: set[int] = set()
    results = []

    for entry in config.monitored_processes:
        found = False
        mem_bytes = 0
        keywords_lower = [k.strip().lower() for k in entry.keyword.split("|")]

        matched_pids: set[int] = set()
        for proc in psutil.process_iter(["pid", "name", "cmdline", "memory_info"]):
            try:
                if proc.pid in global_counted:
                    continue
                cmdline = " ".join(proc.info.get("cmdline") or [])
                cmdline_lower = cmdline.lower()
                if any(kw in cmdline_lower for kw in keywords_lower):
                    matched_pids.add(proc.pid)
            except (psutil.NoSuchProcess, psutil.AccessDenied):
                continue

        if matched_pids:
            found = True
            for pid in matched_pids:
                try:
                    p = psutil.Process(pid)
                    if pid not in global_counted:
                        global_counted.add(pid)
                        mem_bytes += p.memory_info().rss
                    for child in p.children(recursive=True):
                        if child.pid not in global_counted:
                            global_counted.add(child.pid)
                            try:
                                mem_bytes += child.memory_info().rss
                            except (psutil.NoSuchProcess, psutil.AccessDenied):
                                pass
                except (psutil.NoSuchProcess, psutil.AccessDenied):
                    continue

        results.append(ProcessInfo(
            name=entry.name, running=found, mem_bytes=mem_bytes,
        ))
    return results


async def _get_docker_process(container_name: str) -> tuple[bool, int]:
    """检查 Docker 容器状态和内存占用。"""
    try:
        proc = await asyncio.create_subprocess_exec(
            "docker", "inspect", "--format", "{{.State.Running}}", container_name,
            stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
        )
        stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=5)
        running = stdout.decode().strip().lower() == "true"
        if not running:
            return False, 0

        proc2 = await asyncio.create_subprocess_exec(
            "docker", "stats", "--no-stream", "--format", "{{.MemUsage}}", container_name,
            stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE,
        )
        stdout2, _ = await asyncio.wait_for(proc2.communicate(), timeout=5)
        mem_str = stdout2.decode().strip().split("/")[0].strip()
        mem_bytes = _parse_docker_mem(mem_str)
        return True, mem_bytes
    except Exception:
        return False, 0


def _parse_docker_mem(s: str) -> int:
    """解析 docker stats 的内存字符串，如 '120.5MiB' -> bytes。"""
    s = s.strip()
    try:
        if s.endswith("GiB"):
            return int(float(s[:-3]) * 1024 ** 3)
        elif s.endswith("MiB"):
            return int(float(s[:-3]) * 1024 ** 2)
        elif s.endswith("KiB"):
            return int(float(s[:-3]) * 1024)
        elif s.endswith("B"):
            return int(float(s[:-1]))
    except (ValueError, IndexError):
        pass
    return 0


async def get_processes() -> list[ProcessInfo]:
    results = _get_processes_psutil()
    for entry in config.docker_processes:
        running, mem_bytes = await _get_docker_process(entry.container)
        results.append(ProcessInfo(name=entry.name, running=running, mem_bytes=mem_bytes))
    return results


# ================= 网络检测 =================

async def _ping_host(host: str, timeout: int = 3) -> NetworkResult:
    try:
        proc = await asyncio.create_subprocess_exec(
            "ping", "-c", "1", "-W", str(timeout), host,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, _ = await asyncio.wait_for(proc.communicate(), timeout=timeout + 2)
        if proc.returncode == 0:
            output = stdout.decode("utf-8", errors="ignore")
            latency = None
            for line in output.splitlines():
                if "time=" in line:
                    try:
                        t_part = line.split("time=")[1].split()[0]
                        latency = float(t_part.replace("ms", ""))
                    except (IndexError, ValueError):
                        pass
                    break
            return NetworkResult(host=host, reachable=True, latency_ms=latency)
        else:
            return NetworkResult(host=host, reachable=False, latency_ms=None)
    except Exception:
        return NetworkResult(host=host, reachable=False, latency_ms=None)


async def get_network() -> list[NetworkResult]:
    results = await asyncio.gather(
        *[_ping_host(host) for host in config.ping_hosts]
    )
    return list(results)


# ================= 汇总 =================

async def collect_server_status() -> ServerStatus:
    resources = get_resources()
    processes = await get_processes()
    network = await get_network()

    return ServerStatus(
        hostname=platform.node(),
        cpu_model=get_cpu_model(),
        cpu_cores=psutil.cpu_count(logical=True) or 0,
        uptime=get_system_uptime(),
        resources=resources,
        processes=processes,
        network=network,
    )

File: plugins/alive_stat/collector.go
package alive_stat

import (
	"context"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/HakuchumuHYX/HakuBot/utils/logging"
	"github.com/shirou/gopsutil/v4/cpu"
	"github.com/shirou/gopsutil/v4/disk"
	"github.com/shirou/gopsutil/v4/host"
	"github.com/shirou/gopsutil/v4/mem"
	"github.com/shirou/gopsutil/v4/process"
)

type resource struct {
	Name        string
	Percent     float64
	Used, Total uint64
	Available   bool
}

type processInfo struct {
	Name        string
	Status      string
	Memory      uint64
	MemoryKnown bool
}

type networkResult struct {
	Host    string
	Status  string
	Latency *float64
}

type serverStatus struct {
	Hostname, CPUModel, Uptime string
	Cores                      int
	Resources                  []resource
	Processes                  []processInfo
	Network                    []networkResult
}

func collect(ctx context.Context, cfg config) serverStatus {
	result := serverStatus{
		Hostname: "N/A", CPUModel: "N/A", Uptime: "N/A",
		Resources: []resource{{Name: "CPU"}, {Name: "Mem"}, {Name: "Swap"}, {Name: "Disk"}},
		Processes: make([]processInfo, len(cfg.MonitoredProcesses)+len(cfg.DockerProcesses)),
		Network:   make([]networkResult, len(cfg.PingHosts)),
	}
	var tasks sync.WaitGroup
	run := func(work func()) {
		tasks.Add(1)
		go func() {
			defer tasks.Done()
			work()
		}()
	}
	run(func() {
		if name, err := os.Hostname(); err == nil {
			result.Hostname = name
		} else {
			report("hostname", err)
		}
		if info, err := cpu.InfoWithContext(ctx); err == nil && len(info) > 0 {
			result.CPUModel = info[0].ModelName
		} else {
			report("CPU model", err)
		}
		if count, err := cpu.CountsWithContext(ctx, true); err == nil {
			result.Cores = count
		} else {
			report("CPU cores", err)
		}
		if uptime, err := host.UptimeWithContext(ctx); err == nil {
			result.Uptime = duration(float64(uptime))
		} else {
			report("system uptime", err)
		}
	})
	run(func() {
		values, err := cpu.PercentWithContext(ctx, 500*time.Millisecond, false)
		if err == nil && len(values) > 0 {
			result.Resources[0].Percent, result.Resources[0].Available = values[0], true
		} else {
			report("CPU usage", err)
		}
	})
	run(func() {
		if value, err := mem.VirtualMemoryWithContext(ctx); err == nil {
			percent := 0.0
			if value.Total > 0 {
				percent = float64(value.Total-min(value.Total, value.Available)) / float64(value.Total) * 100
			}
			// Preserve psutil's used bytes while its percentage is based on available memory.
			used := value.Total - min(value.Total, value.Free+value.Buffers+value.Cached)
			result.Resources[1] = resource{"Mem", percent, used, value.Total, true}
		} else {
			report("memory", err)
		}
		if value, err := mem.SwapMemoryWithContext(ctx); err == nil {
			result.Resources[2] = resource{"Swap", value.UsedPercent, value.Used, value.Total, true}
		} else {
			report("swap", err)
		}
		if value, err := disk.UsageWithContext(ctx, "/"); err == nil {
			result.Resources[3] = resource{"Disk", value.UsedPercent, value.Used, value.Total, true}
		} else {
			report("disk", err)
		}
	})
	run(func() { copy(result.Processes, collectProcesses(ctx, cfg.MonitoredProcesses)) })
	for i, item := range cfg.DockerProcesses {
		run(func() { result.Processes[len(cfg.MonitoredProcesses)+i] = dockerProcess(ctx, item) })
	}
	for i, name := range cfg.PingHosts {
		run(func() { result.Network[i] = pingHost(ctx, name) })
	}
	tasks.Wait()
	return result
}

func report(part string, err error) {
	if err != nil && !errors.Is(err, context.Canceled) {
		logging.Module("alive_stat").WithField("collector", part).WithError(err).Warn("采集不可用")
	}
}

type processSnapshot struct {
	pid, parent int32
	command     string
	memory      uint64
	memoryKnown bool
}

func collectProcesses(ctx context.Context, entries []processEntry) []processInfo {
	result := make([]processInfo, len(entries))
	for i, entry := range entries {
		result[i] = processInfo{Name: entry.Name, Status: "Stopped"}
	}
	if len(entries) == 0 {
		return result
	}
	all, err := process.ProcessesWithContext(ctx)
	if err != nil {
		report("processes", err)
		for i := range result {
			result[i].Status = "Unknown"
		}
		return result
	}
	rows := make(map[int32]processSnapshot)
	children := make(map[int32][]int32)
	incomplete := false
	for _, proc := range all {
		if ctx.Err() != nil {
			return result
		}
		parent, parentErr := proc.PpidWithContext(ctx)
		command, commandErr := proc.CmdlineWithContext(ctx)
		if errors.Is(parentErr, os.ErrNotExist) || errors.Is(commandErr, os.ErrNotExist) {
			continue
		}
		if parentErr != nil || commandErr != nil {
			incomplete = true
		}
		row := processSnapshot{pid: proc.Pid, parent: parent, command: strings.ToLower(command)}
		if usage, err := proc.MemoryInfoWithContext(ctx); err == nil {
			row.memory, row.memoryKnown = usage.RSS, true
		}
		rows[row.pid] = row
		children[parent] = append(children[parent], row.pid)
	}
	counted := make(map[int32]bool)
	for i, entry := range entries {
		var roots []int32
		for pid, row := range rows {
			if counted[pid] {
				continue
			}
			for _, keyword := range strings.Split(entry.Keyword, "|") {
				keyword = strings.ToLower(strings.TrimSpace(keyword))
				if keyword != "" && strings.Contains(row.command, keyword) {
					roots = append(roots, pid)
					break
				}
			}
		}
		if len(roots) == 0 {
			if incomplete {
				result[i].Status = "Unknown"
			}
			continue
		}
		result[i].Status, result[i].MemoryKnown = "Running", true
		for len(roots) > 0 {
			pid := roots[len(roots)-1]
			roots = roots[:len(roots)-1]
			if counted[pid] {
				continue
			}
			counted[pid] = true
			row := rows[pid]
			result[i].Memory += row.memory
			result[i].MemoryKnown = result[i].MemoryKnown && row.memoryKnown
			roots = append(roots, children[pid]...)
		}
	}
	if incomplete {
		report("processes", fmt.Errorf("some process metadata could not be read"))
	}
	return result
}

func commandOutput(ctx context.Context, timeout time.Duration, name string, args ...string) ([]byte, error) {
	ctx, cancel := context.WithTimeout(ctx, timeout)
	defer cancel()
	cmd := exec.CommandContext(ctx, name, args...)
	cmd.Env = append(os.Environ(), "LC_ALL=C")
	output, err := cmd.Output()
	if ctx.Err() != nil {
		return nil, ctx.Err()
	}
	return output, err
}

func dockerProcess(ctx context.Context, entry dockerEntry) processInfo {
	result := processInfo{Name: entry.Name, Status: "Unknown"}
	output, err := commandOutput(ctx, 5*time.Second, "docker", "inspect", "--format", "{{.State.Running}}", entry.Container)
	if err != nil {
		report(entry.Name, err)
		return result
	}
	switch strings.TrimSpace(string(output)) {
	case "false":
		result.Status = "Stopped"
		return result
	case "true":
		result.Status = "Running"
	default:
		report(entry.Name, fmt.Errorf("unexpected Docker running state"))
		return result
	}
	output, err = commandOutput(ctx, 5*time.Second, "docker", "stats", "--no-stream", "--format", "{{.MemUsage}}", entry.Container)
	if err != nil {
		report(entry.Name, err)
		return result
	}
	value := strings.TrimSpace(strings.SplitN(string(output), "/", 2)[0])
	result.Memory, err = dockerBytes(value)
	result.MemoryKnown = err == nil
	report(entry.Name, err)
	return result
}

var memoryPattern = regexp.MustCompile(`^([0-9]+(?:\.[0-9]+)?)\s*(B|KiB|MiB|GiB|TiB|kB|KB|MB|GB|TB)$`)

func dockerBytes(value string) (uint64, error) {
	match := memoryPattern.FindStringSubmatch(value)
	if match == nil {
		return 0, fmt.Errorf("invalid Docker memory value")
	}
	amount, err := strconv.ParseFloat(match[1], 64)
	if err != nil {
		return 0, err
	}
	units := map[string]float64{"B": 1, "KiB": 1 << 10, "MiB": 1 << 20, "GiB": 1 << 30, "TiB": 1 << 40, "kB": 1e3, "KB": 1e3, "MB": 1e6, "GB": 1e9, "TB": 1e12}
	return uint64(amount * units[match[2]]), nil
}

var latencyPattern = regexp.MustCompile(`time[=<]([0-9]+(?:\.[0-9]+)?)`)

func pingHost(ctx context.Context, name string) networkResult {
	result := networkResult{Host: name, Status: "Unknown"}
	output, err := commandOutput(ctx, 5*time.Second, "ping", "-c", "1", "-W", "3", "--", name)
	if err != nil {
		var exit *exec.ExitError
		if errors.As(err, &exit) && exit.ExitCode() == 1 {
			result.Status = "timeout"
		} else {
			report("ping", err)
		}
		return result
	}
	result.Status = "ok"
	if match := latencyPattern.FindSubmatch(output); match != nil {
		if value, err := strconv.ParseFloat(string(match[1]), 64); err == nil {
			result.Latency = &value
		}
	}
	return result
}

func formatBytes(value uint64) string {
	switch {
	case value < 1024:
		return fmt.Sprintf("%dB", value)
	case value < 1<<20:
		return fmt.Sprintf("%.1fK", float64(value)/(1<<10))
	case value < 1<<30:
		return fmt.Sprintf("%.1fM", float64(value)/(1<<20))
	default:
		return fmt.Sprintf("%.1fG", float64(value)/(1<<30))
	}
}
